### PR TITLE
Add support for using view components in rails engine

### PR DESCRIPTION
**Pull request by tkowalewski**
### What are you trying to accomplish?

I'm trying to add support for generating view components in the module (engine) namespace in rails engine.
To do it we need to wrap templates code with `module_namespacing` (provided by rails)

#### How to test

1. Create new rails engine
```
rails plugin new engine_with_test --mountable
```
2. Update gemspec (`engine_with_test.gemspec`)
  Update `spec.homepage`
  Update `spec.summary`
  Update `spec.description`
  Update `spec.metadata["source_code_uri"]`
  Update `spec.metadata["changelog_uri"]`

3. Add view_component to `Gemfile`
```
gem "view_component", github: "tkowalewski/view_component", branch: "component_template_with_module_namespacing"
```

4. Install gems
```
bundle install
```

5. Generate example component
```
bin/rails generate component Example message --preview
```

6. Update component template (`app/components/engine_with_test/example_component.html.erb`)
```
<span><%= @message %></span>
```

7. Uncomment test (`test/components/engine_with_test/example_component_test.rb`)

8. Run tests
```
bin/rails test
```

9. Preview component
```
bin/rails server
```
and go to `http://localhost:3000/rails/view_components`

### What approach did you choose and why?

There is only one solution to deliver view components for rails engine with module (engine) namespace: Wrap templates code with  `module_namespacing` :)

### Anything you want to highlight for special attention from reviewers?

- I had to prepare example rails engine for tests (`test/test_engine`)
- New rake test task (`engine_test`) was added to Rakefile
- ViewComponent configuration has been extended to provide preview paths for rails engine

### CODE REVIEW

**Comment by boardfish** [comment]
👋 Just a quick nudge to say #1308 means you'll need to rebase this branch for checks to pass.

**Comment by tkowalewski** [comment]
@boardfish @Spone

There is problem with tests for `main` release of rails :|
Any plans to add support for rails `8.0.0.alpha`?

**Comment by reeganviljoen** [comment]
@tkowalewski we have fixed the rails main issues, and added support for ruby head as a bonus aswell

**Comment by joelhawksley** [comment]
@tkowalewski would you be up for fixing the failing tests on this PR? It seems like you're pretty close ❤️

**Comment by tkowalewski** [comment]
@joelhawksley Ready :)

File: Rakefile [hunk 1]
```diff
@@ -8,7 +8,13 @@ require "yard/mattr_accessor_handler"
 Rake::TestTask.new(:test) do |t|
   t.libs << "test"
   t.libs << "lib"
-  t.test_files = FileList["test/**/*_test.rb"]
+  t.test_files = FileList["test/sandbox/**/*_test.rb", "test/view_component/**/*_test.rb"]
+end
+
+Rake::TestTask.new(:engine_test) do |t|
+  t.libs << "test/dummy"
```
Comment (joelhawksley) [inline Rakefile:15]:
```suggestion
  t.libs << "test/engine"
```

👋🏻 I have a slight aversion to using the term `dummy`, might you be willing to change the naming to `engine`?

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,10 @@ nav_order: 5
 
 ## main
 
+* Add support for using view components in Rails Engine
```
Comment (joelhawksley) [inline docs/CHANGELOG.md:13]:
```suggestion
* Add Rails engine support to generators.
```

File: lib/rails/generators/rspec/templates/component_spec.rb.tt [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 require "rails_helper"
 
+<% module_namespacing do -%>
```
Comment (boardfish) [inline lib/rails/generators/rspec/templates/component_spec.rb.tt:5]: I'd prefer it here if the module namespacing was on the described class if we can do that – ideally the `RSpec.describe` block should be top-level. It's sometimes easier to wrap a spec file in a module to save having to mention it all the time, but I don't think we should default to that.

File: Rakefile (hunk 1 above)
Comment (tkowalewski) [reply Rakefile:15]: Hm, `engine` is something like keyword for rails I think.
If we want to rename test engine to `engine` - we will have weird naming I think. For eg. test engine will look like:

```
module Engine
  class Engine < ::Rails::Engine
    isolate_namespace Engine
  end
end
```

File: lib/rails/generators/rspec/templates/component_spec.rb.tt (hunk 3 above)
Comment (tkowalewski) [reply lib/rails/generators/rspec/templates/component_spec.rb.tt:5]: Probably we can try:

```diff
diff --git a/lib/rails/generators/rspec/templates/component_spec.rb.tt b/lib/rails/generators/rspec/templates/component_spec.rb.tt
index 9992f946..7eff719f 100644
--- a/lib/rails/generators/rspec/templates/component_spec.rb.tt
+++ b/lib/rails/generators/rspec/templates/component_spec.rb.tt
@@ -2,8 +2,7 @@

 require "rails_helper"

-<% module_namespacing do -%>
-RSpec.describe <%= class_name %>Component, type: :component do
+RSpec.describe <%= namespaced? ? "#{namespace.name}::" : '' %><%= class_name %>Component, type: :component do
   pending "add some examples to (or delete) #{__FILE__}"

   # it "renders something useful" do
@@ -14,4 +13,3 @@ RSpec.describe <%= class_name %>Component, type: :component do
   #   )
   # end
 end
-<% end -%>
```

Comment (tkowalewski) [reply lib/rails/generators/rspec/templates/component_spec.rb.tt]: Updated for test unit and rspec generator

Comment (boardfish) [reply lib/rails/generators/rspec/templates/component_spec.rb.tt]: Awesome, thank you! 🎉

File: Rakefile (hunk 1 above)
Comment (boardfish) [reply Rakefile:15]: Maybe `test_engine`?

Comment (joelhawksley) [reply Rakefile:15]: Or `example`?

**Review by joelhawksley — APPROVED** [review]
Thank you! ❤️

**Pull request merged**